### DISCUSSION TITLE
How to do basic encryption with a public PGP key

### DISCUSSION THREAD

**Discussion by LazaroR94** [pgpainless/pgpainless#359] [Q&A] [answered]
Hello! I'm new to both this library and PGP cryptography as a whole. My question is in regards to encryption: How exactly can I achieve encryption of a stream using a public PGP key shared with me by a trusted partner? I'm trying to follow some examples here: [Encrypt.java](https://github.com/pgpainless/pgpainless/blob/main/pgpainless-core/src/test/java/org/pgpainless/example/Encrypt.java) but I can't seem to make sense of what I need to do in order to get this encryption mechanism to work? I've figured out decryption just fine but it seems I cannot wrap my head around encryption for some reason. All I have available to me is the public PGP key that I was given and I have added that public key ring as a recipient under the `ProducerOptions.encrypt` method so that it passes it through. I am piping both streams as directed in the examples but for some reason the encrypted stream is never really **encrypted**. What am I doing wrong?

**Comment by vanitasvitae** [comment, marked as answer]
Can you share an example code of what you are doing?

If all you want to do is basic encryption / decryption, take a look at [using PGPainless via `pgpainless-sop`](https://pgpainless.readthedocs.io/en/latest/quickstart.html#encrypt-a-message), which provides the simplest interface for OpenPGP possible :)

Edit: let me know if this solved your issue, or if you have further questions :)

**Reply by LazaroR94** [reply]
Sorry for the late reply! But this is more or less what I am doing to achieve the most basic form of encryption with my partner's public PGP key that was shared with me (Kotlin):
```
PGPainless.encryptAndOrSign()
            .onOutputStream(outputStream)
            .withOptions(
                ProducerOptions.encrypt(
                    EncryptionOptions.encryptCommunications()
                        .addRecipient(certificate)
                )
            )
```

And this is how I am extracting the public PGP certificate that I then later pass to the above code as a **recipient**:
`PGPainless.readKeyRing().publicKeyRing(FileInputStream(pgpPublicKeyLocation))`

**Reply by vanitasvitae** [reply]
The setup looks okay so far. Could you try using another recipient key (e.g. a freshly generated one) to see if maybe something is wrong with your partners key?
If the error still persists with a new key, I would kindly ask you to create a complete example class for me, so I can debug on my end :)

**Reply by LazaroR94** [reply]
So the public PGP key is generated by our enterprise SFT team and I have to use their public key because their private key is stored in the Vault so I cannot create my own and then give it to them. This is how that key is configured:
<img width="617" alt="image" src="https://user-images.githubusercontent.com/98285073/218784729-ed659231-d575-4df5-a84a-4ba79395819b.png">

**Reply by vanitasvitae** [reply]
> So the public PGP key is generated by our enterprise SFT team and I have to use their public key because their private key is stored in the Vault so I cannot create my own and then give it to them.

I understand that. So to figure out whether there is a bug in PGPainless, if the error is on your end or if the key of your partner is somehow incompatible with PGPainless, it would be helpful if you could create a minimal test case for me with a freshly generated key, so I can reproduce the issue on my machine.

**Reply by LazaroR94** [reply]
Sure, let me see if I can create a test class for you (may be some time as I am busy with some other things at the moment unfortunately).

**Reply by vanitasvitae** [reply]
You could also use the [Sequoia Keyring Linter](https://crates.io/crates/sequoia-keyring-linter) to check your partners key for known issues (e.g. SHA-1 used in self signatures, which PGPainless rejects since first of February this year).

**Reply by LazaroR94** [reply]
I don't believe the certificate is signed as per my contact's details. this is the error I seem to get whenever I use their certificate:
`org.pgpainless.exception.KeyException$UnacceptableSelfSignatureException: Key A68D71AD6F17FB1EA0FE044E9B7130FD0594F4CE does not have a valid/acceptable signature to derive an expiration date from.`

**Reply by vanitasvitae** [reply]
The certificate is not well-formed then.
[RFC4880 requires](https://www.rfc-editor.org/rfc/rfc4880#section-12.1) the certificate to have a self-signature, which appears to be missing on this certificate.
Either the key does not have a self-signature at all (For example V3 keys did not require those afaik, but V3 keys are not supported by PGPainless), or it has a self-signature made using an unacceptable hash algorithm like SHA-1.

If the later is the case, you could change PGPainless' default hash algorithm policy to accept weaker algorithms (see [this documentation](https://pgpainless.readthedocs.io/en/latest/quickstart.html#weak-keys-and-broken-algorithms) and [this example](https://github.com/pgpainless/pgpainless/blob/main/pgpainless-core/src/test/java/org/pgpainless/example/ManagePolicy.java#L67-L104)), but that's not recommended in the long run.

Edit: 
> I don't believe the certificate is signed as per my contact's details. this is the error I seem to get whenever I use their certificate:
org.pgpainless.exception.KeyException$UnacceptableSelfSignatureException: Key A68D71AD6F17FB1EA0FE044E9B7130FD0594F4CE does not have a valid/acceptable signature to derive an expiration date from.

Btw. this information would have been useful earlier :P

**Reply by LazaroR94** [reply]
Sorry about that, I realize I should have posted that to begin with haha - would have made the clarification easier. Let me see if I can loop-back with my contact to see how exactly this certificate was set up and I'll get back to you.

P.S. Thank you so much for your prompt replies!

**Reply by vanitasvitae** [reply]
No problem :)

**Reply by LazaroR94** [reply]
Okay so it probably is because it doesn't have a self-signature then as the hashing algorithm appears to be SHA-256. I suppose this is something I can do on my end or does the certificate need to be signed prior to generation?

**Reply by vanitasvitae** [reply]
The signature needs to be made by the secret key over its public key. It is possible to retrospectively self-sign the key, but how thats done depends on the software.
I suspect the key was generated using some legacy software. Recent software would generate the self-signature(s) right away when generating the key (e.g. PGPainless does).

Edit: By the looks of it, the screenshot you showed shows an expiration date, so I suspect the key is infact self-signed, but my suspicion is that the signature was created using SHA-1 (or weaker).
You can do `gpg --list-packet /path/to/cert.asc` to inspect your partners certificate and show its packets. That way you should be able to figure out whether the cert has a selfsignature or not.

**Reply by LazaroR94** [reply]
This is the information retrieved with GPG for that certificate:
```
# off=0 ctb=99 tag=6 hlen=3 plen=525
:public key packet:
	version 4, algo 1, created 1674488333, expires 0
	pkey[0]: [4096 bits]
	pkey[1]: [17 bits]
	keyid: 665A0DE9F0517513
# off=528 ctb=b4 tag=13 hlen=2 plen=17
:user ID packet: "SFT generated key"
# off=547 ctb=89 tag=2 hlen=3 plen=561
:signature packet: algo 1, keyid 665A0DE9F0517513
	version 4, created 1674488333, md5len 0, sigclass 0x13
	digest algo 2, begin of digest 6d e2
	hashed subpkt 2 len 4 (sig created 2023-01-23)
	hashed subpkt 9 len 4 (key expires after 4y1d0h0m)
	hashed subpkt 27 len 1 (key flags: 26)
	hashed subpkt 11 len 1 (pref-sym-algos: 3)
	hashed subpkt 22 len 1 (pref-zip-algos: 1)
	hashed subpkt 21 len 1 (pref-hash-algos: 8)
	critical hashed subpkt 4 len 1 (exportable)
	subpkt 16 len 8 (issuer key ID 665A0DE9F0517513)
	data: [4094 bits]
```

**Reply by vanitasvitae** [reply]
```
:signature packet: algo 1, keyid 665A0DE9F0517513
	version 4, created 1674488333, md5len 0, sigclass 0x13
	digest algo 2, begin of digest 6d e2
	hashed subpkt 2 len 4 (sig created 2023-01-23)
	hashed subpkt 9 len 4 (key expires after 4y1d0h0m)
	hashed subpkt 27 len 1 (key flags: 26)
	hashed subpkt 11 len 1 (pref-sym-algos: 3)
	hashed subpkt 22 len 1 (pref-zip-algos: 1)
	hashed subpkt 21 len 1 (pref-hash-algos: 8)
	critical hashed subpkt 4 len 1 (exportable)
	subpkt 16 len 8 (issuer key ID 665A0DE9F0517513)
	data: [4094 bits]
```

`digest algo 2` indicates that the signature was indeed created using SHA-1.
So you should be fine with setting a custom HashAlgorithPolicy allowing SHA-1.

**Reply by LazaroR94** [reply]
Okay so that seemed to work when I added the HashAlgorithmPolicy override to allow SHA-1. I am still curious about the actual output though, for example, when I print the output of my OutputStream after all is said and done it just prints the content of the public key file itself - is this correct or am I doing something wrong? I am ultimately trying to write the encrypted content to a file.

Encryption method:
```
fun encryptPgpFile(inputStream: ByteArrayInputStream): ByteArray {
        return sop.encrypt()
            .withCert(FileInputStream(pgpPublicKeyLocation))
            .plaintext(inputStream)
            .bytes
    }
```

Encryption test method (logic):
```
        val file = File("path/to/un-encrypted/file")
        val inputStream = ByteArrayInputStream(file.readBytes())

        val byteArray = encryptPgpFile(inputStream)
        val stream = ByteArrayInputStream(byteArray)

        IOUtils.copy(stream, FileOutputStream("path/to/enrypted/file"))
```

**Reply by vanitasvitae** [reply]
If you've wired up everything right, the encrypted file should contain some randomish text starting with `-----BEGIN PGP MESSAGE-----`. To verify whether it is encrypted or not, you could try `gpg -d path/to/encrypted/file`, which should fail because you do not have the secret key available.

**Reply by LazaroR94** [reply]
That is correct. I get the correct header and trailer and it does actually fail when I execute the GPG command. So based on this information, it is safe to assume that the file was successfully encrypted? I'm only curious because I've received some PGP encrypted files before from my partner and the encrypted content looks completely different (A bunch of scribble-scrabble random ASCII characters).

**Reply by vanitasvitae** [reply]
Thats because by default PGPainless wraps the encrypted content in an ASCII armor. That way you can safely copy paste encrypted message to an E-Mail for example. If you do not need ASCII armor (it adds a bit of size overhead), you can disable it by adding `.noArmor()` like so:

```
return sop.encrypt()
                .noArmor() 
                .withCert(FileInputStream(pgpPublicKeyLocation))
                .plaintext(inputStream)
                .bytes
```

**Reply by LazaroR94** [reply]
Oh okay that makes more sense now! And out of curiosity, are there any benefits to using `PGPainless.encryptOrSign()` versus `SOPImpl().encrypt()`? I find that SOP seems a bit more straight-forward as far as I am able to tell (I could be wrong).

**Reply by vanitasvitae** [reply]
The API of `sop.encrypt()` is indeed more straight forward and cleaner, so there are less chances for users (e.g. you 😉 ) to go wrong. Internally, it also calls `PGPainless.encryptOrSign()`.

Another benefit of `sop` is, that you can easily swap out the backend for something other than PGPainless. For example, there is [external-sop](https://github.com/pgpainless/sop-java/tree/main/external-sop) which can be used in place of `pgpainless-sop` to delegate the crypto operations to for example [sequoia-pgp](https://sequoia-pgp.org).

See [this blog post](https://blog.jabberhead.tk/2023/01/13/use-any-sop-binary-with-sop-java-and-external-sop/) for more details.
